### ISSUE TITLE
Proposal for new WCAG 2.2 SC 'States discernible' (a more limited aspect of affordance)

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#559]
Just a very first draft of what after today's telco I perceive as a valid addition to WCAG in version 2.2 (though I can forsee the difficulty of nailing this down unambiguously):

**1.4.XX States discernible.** Where user interface components can have different functional states, these can be discerned visually through changes of salient parts. Salient changes can be:
- a change of the label of the user interface component reflecting the state change
- An addition or removal of a salient part of the interface component, for example, a checkmark in a checkbox or a select element, a border, or an underline
- A change of the form of a salient part, for example, a menu icon turning into a close icon when the menu is opened
- A change of the direction of a salient part, for example, a downwards pointing arrow turning into an upwards pointing arrow when a hidden element is disclosed or a sort direction is changed

Note: Text must meet the contrast requirement of 4.5:1 of SC 1.4.3 and salient not-text must meet the contrast requirement of 3:1 of SC 1.4.11.

**Comment by gianwild** [comment]
Another example is the accordion - a plus sign changing to a minus sign.
Also, is it possible to find a more plain-language word for salient? Such as "notable" or "relevant" or "pertinent"?

**Comment by detlevhfischer** [comment]
@gianwild yes, quite. 
I realise this may be at the edge of being more of a usability criterion but I would argue it affects people with disabilities more. Just an example: people using magnification may not see the effect of toggling a control if it is only the change of other things elsewhere on the screen that makes the state change obvious to users. 
A second remark: This could also align with Operable (i.e. in the vain of 2.4.7 'Focus visible').

**Comment by gianwild** [comment]
@detlevhfischer I definitely think it is an accessibility criterion - for example with an accordion that uses glyphicons people who can visually see the page can see the change in the glyphicon, but a screen reader user won't have access to that at all. In a hamburger menu example, having the alternative to the menu icon be "Menu" and then when it changes to a Close cross remaining "Menu" is definitely an accessibility issue. I would argue that it is a failure of 1.1.1: Non-Text Content

**Comment by mbgower** [comment]
Thanks for taking a first crack at this, @detlevhfischer.

I don't think "functional" is defined anywhere (and may not be the best word), so one of the things to do will be to clearly indicate which states for what controls meet that "functional" definition.

I agree there are real challenges tackling this. There's a convention, for example, that Focus indication is additive to an existing shape for many components (thinking specifically of the focus outline that normally expands the footprint) . However many other states  (and components) do not have even this level of common usage.

It may be an idea to list all the states of all common UI components and have a discussion about which ones are most critical to perceive.

**Comment by alastc** [comment]
I think we could draw a line to name/role/value, so if the value of the component changes that must be visible.

Off the top of my head this should apply to:

- "Selected" for checkboxes & radio buttons, tabs (aria-checked/selected?)
- Location of the 'thumb' for a sldier (aria-valuenow?)
- Visible indicators of show/hide controls (aria-expanded?)

Any others?

I'd probably start with something like:
> 1.4.XX States discernible. User interface components that provide a non-text value or status indicator have a contrast ratio of at least 3:1 against adjacent color(s).

**Comment by mraccess77** [comment]
@alastc your text reads more like the current SC where states by themselves are already covered.  What we want covered is the differentiation of specific states from the opposing state.

**Comment by alastc** [comment]
Ach, good point, it's the status that's the object. 

> 1.4.XX States discernible. The non-text value or status indicator of user interface components have a contrast ratio of at least 3:1 against adjacent color(s).

**Comment by awkawk** [comment]
I wonder if we should try to be more specific in order to avoid the issues we have come up against this time.  So instead of referring to "states" we indicate _which_ states:

e.g. The focus state for a user interface component must...

**Comment by alastc** [comment]
We have a bit of a list, so it would be:
The non-text indicators of selected, expanded and value states must...

It feels like there should be a better collective term for things that visually represent a state of something that impacts the rest of the page.  

- Expanded means there's more content elsewhere, 
- Selected is a value submitted (checkbox) or,
- Selected means there's more content elsewhere (tabs), 
- The slider indicates a value submitted.

**Comment by mbgower** [comment]
I think @awkawk is correct that we should nail down the states in the text. I also think we should concentrate on the most 'important' states. I'm also going to suggest we focus on the idea of 'distinctive'. How about:
> 1.4.XX State Distinguishable. The indicators of changes to the checked, expanded, pressed and selected states for user interface components are distinguishable in one or more of the following ways:
>- 3:1 contrast ratio with non-text indicator of prior state
>- alteration of non-text indicator shape[ by at least ___]
>- visible text indicator of the change in state

or we could just make this for non-text state indicators and make an exception where visible text indicates the state.

**Comment by alastc** [comment]
I think the recent updates solve this issue, for things which have a visual indicator they are covered by the SC & understanding doc.

If there is a gap, it is for things which change hue/luminance & maintain contrast, but do not necessarily contrast within the component. E.g. a link that uses a subtle background change. For that we need to define or add to 'visible' in focus-visible.

Any objection to closing this issue?
